### ISSUE TITLE
Some ideas to improve scaling behaviour

### ISSUE THREAD

**Issue by mreineck** [flatironinstitute/finufft#160]
The spreading routines currently seem to do unnecessary copying in 2D and 3D transforms.

As a simple example, imagine a large 2D spreading task which is cut into many different subproblems in `spreadSorted`.
If one of these subproblems contains a few points falling into the last `x` bin of the first row of `y` bins, and after these it also contains points falling into the first `x` bin of the second row of `y` bins, then the temporary buffer allocated for this subproblem is very large, and only a very small portion of it will be touched during the spreading process. Nevertheless
- it has to be allocated and filled with zeros, and
- it needs to be added in full size to the result array after spreading
Both aspects are sub-optimal, but the second is worse, since this has to happen inside an `#pragma omp critical` region, which can cause quite bad performance when the number of threads is large.

In 3D the scenario is similar and can occur both at the end of an `x` row of bins and at the end of an `xy` slab of bins, where the unnecessary work overhead is even larger.

This could be avoided by making sure that a subproblem never extends beyond the end of a row of bins, but I have not yet found an easy way to adjust the code.

Generally, scaling of FINUFFT may be improved by using mutexes or OpenMP locks to get exclusive write access to parts of the array instead of using `#pragma omp critical`; this allows much finer control over what is actually locked, and may prevent unnecessary serialization.

**Comment by ahbarnett** [comment]
Dear Martin,  Thanks for your interesting thoughts!

A) Re your first point about Cartesian bin ordering being a problem:
Maybe, but my hunch is that it is within a factor of 2 of optimal, whatever the density.
Consider "high" density (in 2D, # NU pts per binning-row > maxsubproblemsize): then most subprobs don't split over rows, and when they do it copy-adds a cuboid of about the size of the previous non-splitting cuboids combined. Ie, no more than half the copy-add effort is wasted.
Now consider low density, then the subprob is always already the full row length, so there is not much copy-add effort wasted.
The way to verify this would be to time a spreading task w/ NU pts grouped into clusters of size exactly maxsubproblemsize (1e4), to remove all large-cuboid subproblems.  If this is much faster than random NU pts, then you are right, and something should be improved.

B) Your point re locks/mutexes being faster than ``omp critical``. I don't know how to code this up, since different threads write to different blocks.
But, is the speedup likely to be significant?
Here's some tests:
First, i7 chip:
A 3D task, narrow w=4 kernel, comparing default to setting timing flag=1 which omits the copy-add to final array:
```
(idp) alex@fiona /home/alex/numerics/finufft> perftest/spreadtestnd 3 1e7 1e7 1e-3 2 0
making random data...
spreadinterp 3D, 9.94e+06 U pts, dir=1, tol=0.001: nspread=4
    1e+07 NU pts in 0.561 s 	1.78e+07 pts/s 	1.14e+09 spread pts/s
    rel err in total over grid:      0.000927
making more random NU pts...
spreadinterp 3D, 9.94e+06 U pts, dir=2, tol=0.001: nspread=4
    1e+07 NU pts in 0.435 s 	2.3e+07 pts/s 	1.47e+09 spread pts/s
    max rel err in values at NU pts: 0.00208
(idp) alex@fiona /home/alex/numerics/finufft> perftest/spreadtestnd 3 1e7 1e7 1e-3 2 1
making random data...
spreadinterp 3D, 9.94e+06 U pts, dir=1, tol=0.001: nspread=4
    1e+07 NU pts in 0.469 s 	2.13e+07 pts/s 	1.36e+09 spread pts/s
    rel err in total over grid:      -nan
making more random NU pts...
spreadinterp 3D, 9.94e+06 U pts, dir=2, tol=0.001: nspread=4
    1e+07 NU pts in 0.449 s 	2.23e+07 pts/s 	1.42e+09 spread pts/s
    max rel err in values at NU pts: inf
```
This tells us that in dir=1 (spread), 16% of time is taken by copy-add. So, that could maybe be halved, giving 8% speed-up overall.
I've checked that the ``du0`` malloc takes insignificant time.
Finally, simply removing the ``#pragma omp critical`` speeds up spreading by 8%, so that's a possibly few % gain if a better mutex is found, but I don't know how to do such mutexes - 
it could get complicated since you want subproblems to copy-add nearby in DRAM to reuse cache lines, but now also be non-overlapping to allow mutexes.

On a 12-core xeon at 24 threads, the 16% above becomes 29%, so maybe 14% is available there?
And it seems 22% is spent on waiting for the ``critical``, which is also worse.
Note this is full hyperthreading, and this may explain why dir=1 gets inefficient w/ many threads
(12 threads is actually a bit faster for dir=1 on this machine!)

For a larger width w these changes are less significant.
I also checked that 2D and 1D have less changes than 3D. Ie the above is close to worst-case. 
So you see that the changes are at the few % level for most problems, if the user doesn't request too many threads.
So, it's hard to know how much effort to spend on this.

It seems like the real gain here would be a mutex that allows many threads to be used without a global critical wait to be done (which gets less efficient the more threads there are..)
If you have any pointers for how to do that for subproblems copy-adding to the big grid, let me know!
Even better, if you can contribute code that speeds up the above dir=1 tests for a large # threads, great.

Best, Alex

**Comment by mreineck** [comment]
> Maybe, but my hunch is that it is within a factor of 2 of optimal, whatever the density.

In 3D, it can be a factor of 3. Sure, still a constant factor and not problematic in itself. But the time spent doing this copying is (currently) in a critical OpenMP section, where it is really important to save as much time as possible. As soon as there are enough threads working on a problem that they have to queue for access to that region, overall performance will plummet.

> B) Your point re locks/mutexes being faster than omp critical. I don't know how to code this up, since different threads write to different blocks.

The general idea is to not restrict entry to the code, but only delay execution if data is copied to a part of the array that is already being written to.
Let's assume you are solving a 2D problem: in that case you can create a `vector<omp_lock_t>` with every entry corresponding to exacty one line in the 2D array. Whenever a thread wants to write to some line of the array, it calls `omp_set_lock()` on the corresponding lock. If the lock is already in use by another thread, this call will block until the lock has been released. Once the call has returned, it is safe to write to the line of the array. Afterwards, the lock is released with `omp_unset_lock()`.
This version allows concurrent execution of copy operations that work on disjoint regions of the array, which can help a lot.
Of course this is only necessary once the critical region really becomes a bottleneck, i.e. when a spreading operation becomes noticeably slower than the analogous interpolation operation. But I think this is a situation you are already observing.

> Even better, if you can contribute code that speeds up the above dir=1 tests for a large # threads, great.

I'll see what I can do, but at the moment I'm a bit over-committed unfortunately.

**Comment by blackwer** [comment]
> Let's assume you are solving a 2D problem: in that case you can create a vector<omp_lock_t> with every entry corresponding to exacty one line in the 2D array.
This is a similar suggestion to what I made via private channels. I'll give this and some other ideas a shot, so no worries about your current commitment level.

**Comment by ahbarnett** [comment]
Closing since @blackwer 's fix really helped in the high-thread-count case, as fixed in #166.
